src/entity.ts

```diff
@@ -89,20 +89,73 @@ export namespace entity {
    *
    * @class
    * @param {number|string} value The integer value.
+   * @param {object} typeCastOptions Config for custom `integerValue` cast.
+   * @param {function} [typeCastOptions.integerTypeCastFunction] A custom user
+   *     provided function to convert `integerValue`.
+   * @param {sting|string[]} [typeCastOptions.properties] `Entity` property
+   *     names to be converted using `integerTypeCastFunction`.
    *
    * @example
    * const {Datastore} = require('@google-cloud/datastore');
    * const datastore = new Datastore();
    * const anInt = datastore.int(7);
    */
-  export class Int {
+  export class Int extends Number {
     value: string;
-    constructor(value: number | string) {
+    typeCastFunction?: Function;
+    typeCastProperties?: string[];
+    private _entityPropertyName: string | undefined;
+    constructor(
+      value: number | string | ValueProto,
+      typeCastOptions?: IntegerTypeCastOptions
+    ) {
+      super();
+      this._entityPropertyName =
+        typeof value === 'object' ? value.propertyName : undefined;
+      this.value =
+        typeof value === 'object'
+          ? value.integerValue.toString()
+          : value.toString();
       /**
        * @name Int#value
        * @type {string}
        */
-      this.value = value.toString();
+      if (typeCastOptions) {
+        if (typeof typeCastOptions.integerTypeCastFunction !== 'function') {
+          throw new Error(
+            `integerTypeCastFunction is not a function or was not provided.`
+          );
+        }
+        this.typeCastFunction = typeCastOptions.integerTypeCastFunction;
+        this.typeCastProperties = typeCastOptions.properties
+          ? arrify(typeCastOptions.properties)
+          : undefined;
+      }
+    }
+    // tslint:disable-next-line no-any
+    valueOf(): any {
```

Comment:
We can make this type-safe for typescript users by making this class generic on the return type of `valueOf()`:

```
class Int<T> extends Number ...
```

We'll also be able to make `integerTypeCastFunction` type-safe by making `IntegerTypeCastOptions` generic:

```typescript
export interface IntegerTypeCastOptions<T> {
  integerTypeCastFunction: (value: string) => T;
  properties?: string | string[];
}
```

Reply:
Nevermind folks.. It's not possible to override the return type of `valueOf()` to the generic type since `Int` inherits from `Number`. Typescript really dislike type-hacking..

Reply:
This also means Typescript users would have to coerce type returned by `valueOf()`:

```typescript
// assume integerTypeCastFunction returns a BigInt
const bigInt = myInt.valueOf() as BigInt;
```